java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -188,22 +205,79 @@ public BooleanConstraint shouldNotInverse() {
     return states;
   }
 
-  @CheckForNull
-  private ProgramState checkRelation(BooleanConstraint booleanConstraint, ProgramState programState) {
+  private boolean checkRelation(BooleanConstraint booleanConstraint, ProgramState programState) {
     RelationState relationState = binaryRelation().resolveState(programState.getKnownRelations());
-    if (relationState.rejects(booleanConstraint)) {
-      return null;
+    return !relationState.rejects(booleanConstraint);
+  }
+
+  private List<RelationalSymbolicValue> transitiveRelations(ProgramState programState) {
+    BinaryRelation relation = binaryRelation();
+    return programState.getKnownRelations().stream()
+      .map(r -> r.deduceTransitiveOrSimplified(relation))
+      .filter(Objects::nonNull)
+      .map(RelationalSymbolicValue::binaryRelationToSymbolicValue)
+      .collect(Collectors.toList());
+  }
+
+  private static RelationalSymbolicValue binaryRelationToSymbolicValue(BinaryRelation binaryRelation) {
+    switch (binaryRelation.kind) {
+      case EQUAL:
+      case METHOD_EQUALS:
+      case LESS_THAN:
+        return new RelationalSymbolicValue(binaryRelation.kind, binaryRelation.leftOp, binaryRelation.rightOp);
+      case NOT_EQUAL:
+        return new RelationalSymbolicValue(Kind.NOT_EQUAL, binaryRelation.leftOp, binaryRelation.rightOp);
+      case GREATER_THAN:
+        return new RelationalSymbolicValue(Kind.LESS_THAN, binaryRelation.rightOp, binaryRelation.leftOp);
+      case GREATER_THAN_OR_EQUAL:
+        return new RelationalSymbolicValue(Kind.GREATER_THAN_OR_EQUAL, binaryRelation.leftOp, binaryRelation.rightOp);
+      case LESS_THAN_OR_EQUAL:
+        return new RelationalSymbolicValue(Kind.GREATER_THAN_OR_EQUAL, binaryRelation.rightOp, binaryRelation.leftOp);
+      case NOT_METHOD_EQUALS:
+        return new RelationalSymbolicValue(Kind.NOT_METHOD_EQUALS, binaryRelation.rightOp, binaryRelation.leftOp);
```

Comment:
isn't there something odd with the order of operand here ? if so it can be merged with the first one.